### PR TITLE
Minimum cards in a container

### PR DESCRIPTION
**Pull request by AshCorr**
## What does this change?

 - Removes all unsafe index access from containers by using slices instead of directly referencing indexes.
 - Remvoes some fallbacks that were previously made

## Adding restrictions upstream to gurantee correct number of cards

Following on from a discussion with EdTools about this we identified some problems that might make it difficult if not impossible to gurantee this upstream:

- The config tool is unaware of how many cards will be backfilled from the backfill query, this would only be known at the pressing stage by which time it would be too late for an editor to be warned.
- It might be difficult from a UX perspective to prevent an editor from changing a container to a larger type without first increasing how many cards it holds.

### Falling back to a smaller container

An idea that has been bounced about is the idea of falling back to a smaller container type if theres not enough cards to fill the current container type. This would be great as we wouldn't need to think about fallback design much at all.

The problem with this approach is that we don't have containers types for each size variation. Taking `fixed/small/slow-IV` for example, its only fallback of the same category would be `fixed/small/slow-I`. 

Its even worse with containers such as `fixed/large/slow-XIV` which is the only `fixed/large` container that exists. Maybe it could fall back to `dynamic/slow` but even that only supports 8 cards vs XIV's 14 cards and would be a significant visual change.

### Building fallbacks for each container

Ideally for visual reasons we would like to have fallbacks for every number of cards, but I think this introduces a lot of complexity & confusion into containers that we may want to avoid.

By not falling back it ensures containers remain simple & atleast somewhat true to their original designs. If cards are missing in most cases there will simply be an empty space where a card should have been which is a pretty striking indication that an editor has made a mistake on the front.

### CODE REVIEW

**Comment by mxdvl** [comment]
I absolutely love the idea of using `map` instead of accessing direct members of arrays! I’m not so convinced about the benefit of `SafeArray`, though, but I think these changes can be looked at independently?

File: dotcom-rendering/src/web/components/DynamicFast.tsx [hunk 1]
```diff
@@ -485,9 +485,9 @@ export const DynamicFast = ({
 		firstSliceLayout = 'oneVeryBig';
 		firstSliceCards = groupedTrails.veryBig.slice(0, 1);
 	} else if (groupedTrails.veryBig.length > 1) {
-		if (groupedTrails.veryBig[0].isBoosted) {
+		if (groupedTrails.veryBig[0]?.isBoosted) {
```
Comment (AshCorr) [inline dotcom-rendering/src/web/components/DynamicFast.tsx:488]: This PR was made with `noUncheckedIndexedAccess` enabled, these changes are done in preparation for enabling it across the codebase.

File: dotcom-rendering/src/web/components/DynamicPackage.stories.tsx [hunk 2]
```diff
@@ -281,8 +281,8 @@ Boosted3.story = {
 };
 
 export const Boosted4 = () => {
-	const primary = [...trails].slice(0)[0];
-	const remaining = [...trails].slice(1, 4);
+	const primary = trails[0];
+	const remaining = trails.slice(1, 4);
```
Comment (mxdvl) [inline dotcom-rendering/src/web/components/DynamicPackage.stories.tsx:285]: What do you think of the following?

```suggestion
	const [primary, ...remaining] = trails.slice(0, 4);
```

File: dotcom-rendering/src/web/components/FixedSmallSlowIII.tsx [hunk 3]
```diff
@@ -15,11 +15,11 @@ export const FixedSmallSlowIII = ({
 	containerPalette,
 	showAge,
 }: Props) => {
-	const smallTrails = trails.slice(0, 3);
+	const primary = trails.slice(0, 3);
 
 	return (
 		<UL direction="row">
-			{smallTrails.map((trail, index) => {
+			{primary.map((trail, index) => {
```
Comment (mxdvl) [inline dotcom-rendering/src/web/components/FixedSmallSlowIII.tsx:22]: We might as well inline that directly?

```suggestion
			{trails.slice(0, 3).map((trail, index) => {
```

File: dotcom-rendering/src/web/lib/dynamicSlices.tsx [hunk 4]
```diff
@@ -34,24 +34,30 @@ export const Card100 = ({
...
-			/>
-		</LI>
-	</UL>
-);
+}) => {
+	const primaries = cards.slice(0, 1);
+
+	return (
+		<UL>
+			{primaries.map((trail) => (
```
Comment (mxdvl) [inline dotcom-rendering/src/web/lib/dynamicSlices.tsx:42]: I would probably opt for an inline slice, here, or keep “primaries” singular.

```suggestion
			{card.slice(0, 1).map((trail) => (
```

File: dotcom-rendering/src/web/lib/dynamicSlices.tsx [hunk 5]
```diff
@@ -145,31 +155,40 @@ export const Card25_Card75 = ({
...
-				imagePosition="left"
-				imageSize="large"
-				imagePositionOnMobile="top"
-			/>
-		</LI>
-	</UL>
-);
+}) => {
+	const primaries = cards.slice(0, 1);
+	const secondaries = cards.slice(1, 2);
```
Comment (mxdvl) [inline dotcom-rendering/src/web/lib/dynamicSlices.tsx:160]:
```suggestion
	const card25 = cards.slice(0, 1);
	const card75 = cards.slice(1, 2);
```

File: dotcom-rendering/src/web/lib/dynamicSlices.tsx [hunk 6]
```diff
@@ -186,30 +205,39 @@ export const Card66_Card33 = ({
...
-				showAge={showAge}
-				trailText={cards[1].trailText}
-				imagePositionOnMobile="left"
-			/>
-		</LI>
-	</UL>
-);
+}) => {
+	const primaries = cards.slice(0, 1);
+	const secondaries = cards.slice(1, 2);
```
Comment (mxdvl) [inline dotcom-rendering/src/web/lib/dynamicSlices.tsx:210]:
```suggestion
	const card66 = cards.slice(0, 1);
	const card33 = cards.slice(1, 2);
```

File: dotcom-rendering/src/web/lib/dynamicSlices.tsx [hunk 7]
```diff
@@ -226,56 +254,51 @@ export const Card50_Card25_Card25 = ({
...
-				supportingContent={
-					cards[2].trailText ? undefined : cards[2].supportingContent
-				}
-			/>
-		</LI>
-	</UL>
-);
+}) => {
+	const primaries = cards.slice(0, 1);
+	const secondaries = cards.slice(1, 3);
```
Comment (mxdvl) [inline dotcom-rendering/src/web/lib/dynamicSlices.tsx:259]:
```suggestion
	const card50 = cards.slice(0, 1);
	const cards25 = cards.slice(1, 3);
```

**Review by mxdvl — COMMENTED** [review]
This is great!

Some comments around extracting variables. I would either opt for inline directly, or keep the names of the slices aligned with the components.

```tsx
// For example
const Card50_Card25_Card25 = ({ cards } : Props) => {
  const card50 = cards.slice(0, 1);
  const cards25 = cards.slice(1, 3);
  return ( <>/* … */</> )
}
```

File: dotcom-rendering/src/web/components/DynamicPackage.stories.tsx (hunk 2 above)
Comment (AshCorr) [reply dotcom-rendering/src/web/components/DynamicPackage.stories.tsx:285]: Unfortunately the `trails` array we use in Stories has a guaranteed size, but if you try to apply a function such as slice the array returned won't have the same size gurantees

File: dotcom-rendering/src/web/components/FixedSmallSlowIII.tsx (hunk 3 above)
Comment (AshCorr) [reply dotcom-rendering/src/web/components/FixedSmallSlowIII.tsx:22]: I'm somewhat of a fan of keeping slices explicit at the start of a container, I feel like it makes it clearer how a container is using cards!

File: dotcom-rendering/src/web/lib/dynamicSlices.tsx [hunk 8]
```diff
@@ -73,11 +73,11 @@ export const Card50_Card50 = ({
 	containerPalette?: DCRContainerPalette;
 	showAge?: boolean;
 }) => {
-	const primaries = cards.slice(0, 2);
+	const card50 = cards.slice(0, 2);
```
Comment (mxdvl) [inline dotcom-rendering/src/web/lib/dynamicSlices.tsx:76]: Plural?

```suggestion
	const cards50 = cards.slice(0, 2);
```

**Comment by mxdvl** [comment]
> Need to use keys.

The webUrl of a card should be pretty unique 😁

File: dotcom-rendering/src/web/components/NavList.tsx [hunk 9]
```diff
@@ -53,7 +53,7 @@ export const NavList = ({ trails, containerPalette }: Props) => {
 	return (
 		<ul css={ulStyles}>
 			{trails.map((trail) => (
-				<li css={liStyles(containerOverrides)}>
```
Comment (AshCorr) [inline dotcom-rendering/src/web/components/NavList.tsx:56]: Unrelated to this PR but I missed adding `key` to NavList when I created it.

File: dotcom-rendering/src/web/components/DynamicSlowMPU.tsx [hunk 10]
```diff
@@ -30,53 +30,45 @@ const Card33_ColumnOfThree33_Ad33 = ({
...
-						containerPalette={containerPalette}
-						showAge={showAge}
-						imageUrl={undefined}
-						headlineSize="small"
-					/>
-				</LI>
-				<LI padSides={true}>
+}) => {
+	const primary = cards.slice(0, 1);
+	const secondary = cards.slice(1, 4);
```
Comment (mxdvl) [inline dotcom-rendering/src/web/components/DynamicSlowMPU.tsx:35]:
```ts
	const card33 = cards.slice(0, 1);
	const cards33 = cards.slice(1, 4);
```

File: dotcom-rendering/src/web/components/DynamicSlowMPU.tsx [hunk 11]
```diff
@@ -147,53 +127,51 @@ const Card50_Card25_Card25 = ({
...
-				supportingContent={
-					cards[2].trailText ? undefined : cards[2].supportingContent
-				}
-			/>
-		</LI>
-	</UL>
-);
+}) => {
+	const primary = cards.slice(0, 1);
+	const secondary = cards.slice(1, 3);
```
Comment (mxdvl) [inline dotcom-rendering/src/web/components/DynamicSlowMPU.tsx:132]:
```ts
	const card50 = cards.slice(0, 1);
	const cards25 = cards.slice(1, 3);
```

File: dotcom-rendering/src/web/components/DynamicSlowMPU.tsx [hunk 12]
```diff
@@ -93,46 +85,34 @@ const ColumnOfThree50_Ad50 = ({
...
-		</LI>
-		<LI percentage="50%">
-			<Hide until="tablet">
-				<AdSlot position="inline" index={index} />
-			</Hide>
-		</LI>
-	</UL>
-);
+}) => {
+	const primary = cards.slice(0, 3);
```
Comment (mxdvl) [inline dotcom-rendering/src/web/components/DynamicSlowMPU.tsx:89]:
```ts
	const cards50 = cards.slice(0, 3);
```

File: dotcom-rendering/src/web/components/FixedLargeSlowXIV.tsx [hunk 13]
```diff
@@ -16,38 +16,48 @@ export const FixedLargeSlowXIV = ({
...
 	showAge,
 }: Props) => {
-	const primary = trails[0];
-	const secondary = trails[1];
-	const secondSlice = trails.slice(2, 6);
-	const thirdSlice = trails.slice(6, 14);
+	const firstSlice75 = trails.slice(0, 1);
+	const firstSlice25 = trails.slice(1, 2);
+	const secondSlice25 = trails.slice(2, 6);
+	const thirdSlice25 = trails.slice(6, 14);
```
Comment (mxdvl) [inline dotcom-rendering/src/web/components/FixedLargeSlowXIV.tsx:22]: Not sure if we need to rename the second and third slice, but I like the `firstSlice75` and `firstSlice25`

File: dotcom-rendering/src/web/components/DynamicPackage.tsx [hunk 14]
```diff
@@ -82,31 +82,42 @@ const Card75_Card25 = ({
 	showAge?: boolean;
 }) => {
 	if (cards.length < 2) return null;
+	const primary = cards.slice(0, 1);
+	const secondary = cards.slice(1, 2);
```
[the remark this answers is not in the document]
Comment (AshCorr) [reply dotcom-rendering/src/web/components/DynamicPackage.tsx:86]: Dang I thought I caught them all! thanks!

File: dotcom-rendering/src/web/components/DynamicPackage.tsx [hunk 15]
```diff
@@ -338,23 +349,27 @@ const Card75_ColumnOfCards25 = ({
 	containerPalette?: DCRContainerPalette;
 	showAge?: boolean;
 }) => {
-	const [primary, ...remaining] = cards;
+	const primary = cards.slice(0, 1);
+	const remaining = cards.slice(1);
```
[the remark this answers is not in the document]
Comment (AshCorr) [reply dotcom-rendering/src/web/components/DynamicPackage.tsx:353]: You would imagine there would be wouldn't you... hmm... I think we can probably upper bound it yea, its used by the "threeOrFourStandards" layout which would imply max 4 cards are going into it..

File: dotcom-rendering/src/web/components/FixedLargeSlowXIV.tsx (hunk 13 above)
Comment (AshCorr) [reply dotcom-rendering/src/web/components/FixedLargeSlowXIV.tsx:22]: I think its still useful to a degree, by looking at the top of the container you can easily tell that the 2nd row is composed of 4 25% cards

**Comment by bryophyta** [comment]
I wasn't able to go through all the details in time, but this looks great!

If we don't have it documented somewhere already, I'm wondering if there's a way of making clear that the `[].slice().map()` pattern is the preferred pattern for dealing with lists of cards? Maybe if/when we enable `noUncheckedIndexAccess` it won't matter too much, but in the meantime it could be helpful, if we can think of a good way of surfacing this? (Not super important though, I don't think.)

Great work!

**Comment by mxdvl** [comment]
I think the idea is to enable the flag shortly!

**Pull request merged**